### ISSUE TITLE
Deleted environments staying on project dashboard

### ISSUE THREAD

**Issue by a-fro** [opendevshop/devshop#73]
Hi team Devshop! We've been using devshop since June 2015 for testing features of pull requests for our team. The github pr integration is awesome! However, one issue we've had over time is that deleted environments staying on the dashboard page, and with more than 70 PR's over the course of the past year, the page became really cluttered. 

Recently, we reinstalled with devshop 0.8.1 and started fresh. However, we're still having a couple of issues with environments. I'll describe both here, but please let me know if I should create a separate issue for each.

Here's an example of a deleted environment that is still on the page.

![screen shot 2016-02-25 at 8 13 01 am](https://cloud.githubusercontent.com/assets/1293063/13323064/a82d6106-dba5-11e5-9c99-edea4a509ac2.png)

 Looking at line 111 of nodes.inc, I see

```
// Don't load the environment if the site and/or platform is deleted.
if ($environment->site_status == HOSTING_SITE_DELETED && $environment->platform_status == HOSTING_PLATFORM_DELETED) {
  continue;
}
```

I'm guessing that is the culprit. Digging a bit, I see the node status is 1 for that platform in the nodes table, and the status for that row in the hosting_platform table is also 1. That sits right above another row in the hosting_platform table which is -2. Which may be related to our 2nd issue. 

Platforms fail to automatically delete successfully when a PR is merged or closed. I tried to reproduce this issue, but ran into another (possibly related) issue with the [clone task](https://github.com/opendevshop/devshop/issues/72). The message is something akin to "There are still active sites on this platform". I'm guessing that could be because our git repo master branch (and all subsequent pull requests) include a couple of site folders for our Acquia dev and test environments. The Acquia-specific settings.php files have been force added there, because it enables us to deploy any branch to dev or test on Acquia by just pushing a branch to our Acquia remote. Before devshop, we were testing pull requests on Acquia, but found it too limiting to have only 2 environments. 

If I ssh into the devshop server and delete those site folders (dev.acqui-domain.com), I am able to retry the failed delete task and it succeeds. This is the exact scenario which led to the screenshot above. Is the intention that **any** settings.php file should trigger a failure of environment deletion, or should it just check the settings.php file related to the current domain (ie. branch-[prname].project.devshophost.com)?

**Comment by jonpugh** [comment]
I think this is tied directly to your other issue with the extra github webhook: #72

The trigger for a platform delete is in a POST HOSTING TASK hook for "site delete".  If the platform is found and has no other sites, a delete for the platform is triggered.

I think what you are seeing is that hook not finding the platform  to delete it.

Please setup a new PR environment without the double webhook and give it a test to see if you still have that problem.

Thanks!

**Comment by a-fro** [comment]
@jonpugh unfortunately (or fortunately, depending on your perspective), I think I've found another bug. If there are existing settings.php files in the sites directory, devshop (aegir?) tries adds them as sites and tries to install them. The task fails with

     Calling hook drush_devshop_provision_pre_provision_devshop_deploy
     provision-devshop-deploy must be run on a site context.
Those "sites" don't actually exist in the environment, but they are in the nodes table. When devshop eventually tries to clean up the environment (PR closed), the error is `Existing sites were found on this platform. These sites will need to be deleted before this platform can be deleted.`. If you remove the pre-existing settings.php files and retry the delete task, it succeeds. And the screenshot from above goes away from the dashboard! 

Therefore, I think the bug is two parts: 1) the auto-creation of the pre-existing sites to begin with, and 2) the failure to delete an environment if any settings.php file is found in the sites folder. I did try removing the site nodes from the admin/content/node page, but that didn't work, which is why I'm describing this as a two-part bug. Thanks!

**Comment by jonpugh** [comment]
> there are existing settings.php files in the sites directory, devshop (aegir?) tries adds them as sites and tries to install them

Arggg yes... This is an aegir "feature" that should not exist... I'm not sure yet how we can handle this, let me think on it.

> Those "sites" don't actually exist in the environment, but they are in the nodes table. When devshop eventually tries to clean up the environment (PR closed), the error is Existing sites were found on this platform. These sites will need to be deleted before this platform can be deleted.. If you remove the pre-existing settings.php files and retry the delete task, it succeeds. And the screenshot from above goes away from the dashboard!

Yeah, this too is an aegir thing... it doesn't even need to be active, if there is a settings.php file, it will stop removing things.  

We can figure out how to override these behaviors. I am an aegir maintainer, but it can still be hard how to track it down and fix strange things like this.

What I recommend is we lookup these rogue troublemaking files and add to the new "environment warnings" section in the environments widget.  That way the user is told that their codebase is incompatible with aegir/devshop.

It is very hard to support any possible files and folders layout. Repos might have multisite folders in them.  We simply have to try and detect potential problems and warn users, informing them of the solution, which might be "remove these files from your git repo."

**Comment by jonpugh** [comment]
> 2) the failure to delete an environment if any settings.php file is found in the sites folder.

This is aegir playing it safe.  It is not a bad idea to pause on a deletion if there is a settings.php file.  Perhaps we can get the aegir team to agree that if that settings.php file cannot be bootstrapped, *then* it is safe to delete?

The best thing we can do in devshop is to improve that error message and provide guidance on recourse.  In fact I might make that change upstream in aegir.

**Comment by jonpugh** [comment]
>  I did try removing the site nodes from the admin/content/node page, but that didn't work.

Don't do this. Ever. Unless you go through /var/aegir/config/ and delete the right config files, delete the source code, and the drush aliases manually.

**Comment by a-fro** [comment]
> Don't do this. Ever. Unless you go through /var/aegir/config/ and delete the right config files, delete the source code, and the drush aliases manually.

Yeah, I learned that the hard way. Apache wouldn't restart afterwards until I removed the offending vhost files. Given that these are non-existing platforms, the source code wasn't an issue, though the drush aliases might come back to bite us. No issues thus far, however.

Given that the site creation/deletion issue is related to aegir, I think that the best solution for our use-case is to remove the Acquia-specific settings.php files from the repository. I think we can get the same functionality on Acquia with a pre-push git hook that checks to see which remote is being pushed to, and adding the necessary settings.php file if it's the Acquia remote.
